fix(ivy): support for multiple ICU expressions in the same i18n block

ocombe opened the pull request:
There were two issues with multiple ICU expressions in the same i18n block:
- the regexp that was used to parse the text wasn't able to handle multiple ICU expressions, I've replaced it with parsing the text and searching for brackets (which is what we ended up doing in the end anyway)
- we allocate node indexes for nodes generated by the ICU expressions which increases the expando value, but we would create the nodes for those cases during the update phase. In the mean time we would create some nodes during the creation phase (comment nodes for ICU expressions, text nodes, ...) with an auto increment index. This means that any node created after an ICU expression would get the following index value, but the ICU case nodes expected to use the same index as well... There was a mismatch between the auto generated index, and the expected index which was causing problems when we needed to select those nodes for updates later on. To fix it, I've added the expected node index to the list of mutate codes that we generate, and we do not use an auto increment value anymore.


FW-905 #resolve

mary-poppins commented:
You can preview c53424b at https://pr28083-c53424b.ngbuilds.io/.

AndrewKushnir reviewed (approved):
Looks good to me.

It'd be great if @mhevery or @kara have a look as well.

Thank you.

ocombe committed:
fix(ivy): support for multiple ICU expressions in the same i18n block

There were two issues with multiple ICU expressions in the same i18n block:
- the regexp that was used to parse the text wasn't able to handle multiple ICU expressions, I've replaced it with parsing the text and searching for brackets (which is what we ended up doing in the end anyway)
- we allocate node indexes for nodes generated by the ICU expressions which increases the expando value, but we would create the nodes for those cases during the update phase. In the mean time we would create some nodes during the creation phase (comment nodes for ICU expressions, text nodes, ...) with an auto increment index. This means that any node created after an ICU expression would get the following index value, but the ICU case nodes expected to use the same index as well... There was a mismatch between the auto generated index, and the expected index which was causing problems when we needed to select those nodes for updates later on. To fix it, I've added the expected node index to the list of mutate codes that we generate, and we do not use an auto increment value anymore.

FW-905 #resolve

mary-poppins commented:
You can preview c2d496f at https://pr28083-c2d496f.ngbuilds.io/.

mhevery commented:
http://test/OCL:229307881:BASE:229307905:1547528246221:6725d55a

AndrewKushnir commented:
@ocombe it looks like PR was approved and it's "green" (CirlceCI and g3 checks passed). Please set "merge" label once you're ready. Thank you.
